File: 动态规划/583.两个字符串的删除操作.py
# @before-stub-for-debug-begin
from python3problem583 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode.cn id=583 lang=python3
#
# [583] 两个字符串的删除操作
#

# @lc code=start
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:      
        len1=len(word1)
        len2=len(word2)
        dp=[[0 for j in range(len2+1)]for i in range(len1+1)]
        #初始化
        for j in range(len2+1):
            dp[0][j]=j
        for i in range(len1+1):
            dp[i][0]=i
        for i in range(1,len1+1):
            for j in range(1,len2+1):
                if word1[i-1]==word2[j-1]:#有一个移位
                    dp[i][j]=dp[i-1][j-1]
                else:
                    dp[i][j]=min(dp[i-1][j],dp[i][j-1])+1

        return dp[len1][len2]                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
# @lc code=end

File: 代码随想录/动态规划/583.两个字符串的删除操作.cpp
/*
 * @lc app=leetcode.cn id=583 lang=cpp
 *
 * [583] 两个字符串的删除操作
 */

// @lc code=start
#include <vector>
#include <string>
#include<limits.h>
using namespace std;
class Solution
{
public:
    int minDistance(string word1, string word2)
    {
        int word1Len = word1.size();
        int word2Len = word2.size();
        vector<vector<int>> dp(word1Len + 1, vector<int>(word2Len + 1, INT_MAX));
        // dp[i][j]代表word1(0~i-1)和word2(0~j-1)相等需要删除的最小字符串个数
        for (int i = 1; i < word1Len+1; i++)
        {
            dp[i][0] = i;
        }
        for (int j = 1; j < word2Len+1; j++)
        {
            dp[0][j] = j;
        }
        dp[0][0] = 0;
        for (int i = 1; i < word1Len+1; i++)
        {
            for (int j = 1; j < word2Len+1; j++)
            {
                if (word1[i - 1] == word2[j - 1])
                {
                    dp[i][j]=dp[i-1][j-1];
                }
                else
                {
                    //字符不相等，需要删除其中的一个
                    //选择删除dp[i-1][j],dp[i][j-1]中较大的那一个
                    dp[i][j]=min(dp[i-1][j],dp[i][j-1])+1; 
                }
            }
        }
        return dp[word1Len][word2Len];
    }
};
// @lc code=end
